### ISSUE TITLE
Suggestions for the Node system

### ISSUE THREAD

**Issue by ismslv** [godotengine/godot#28915]
### History
I've read many threads here where people argue about the nodes system,   
compare it with components and so on. They suggest removing scripts   
from nodes or adding multiple scripts. I like the Godot architecture logic   
except several small things, which can be corrected with minimal effort   
to try to resolve inconsistencies.

### Suggestions
I have found two inconsistencies, which for me do not seem logical   
and seem to mess with otherwise logical architecture.   

#### 1. Automatic Nodes in list
- CS scripts extending Nodes should appear automatically as GDScript ones do, or
- GDScript ones should not do it

#### Description
I used only C# for some time and then tried GDScript. I created a custom node type, imported as addon. Then I suddenly saw that this node appears twice in the list: as imported custom one and as automatic one, and I could not understand why I have this second - the same - node type (because with c# I have never seen it). It looked like a strange bug for me. So, to solve this inconsistency, either both languages should do the same (__of course, c# after building__), or neither should do it at all (it is just a shortcut to a node with a script).

#### 2. Custom Nodes
- Custom nodes should be packed

#### Description
With this I mean that custom nodes, defined with ```add_custom_type```, should not have a script attached, because the same could be accomplished with just a script, and using a plugin architecture adds only an icon (even no tree stacking, as I describe in #28890 ). It would be much better and logical, if custom nodes defined in plugins, with all that additional configuration and scripts, would be truly custom: a node with a custom script built-in and an empty script field exposed to add one more script - that is, like the built-in nodes do. Otherwise, it is not logical that ```add_custom_type``` does not really add a custom type, because everywhere inheritance demands only "true" types like Node or Spacial (#28890 and #28908), and "not true" just-scripts can be created without any plugin system.

#### Why not
Of course, I don't mean that addons should compile to native or c++ core scripts as we write them. I mean that custom nodes should have two "script" fields, one of them hidden - that should be much simpler to implement.

**Comment by silkentrance** [comment]
I second that as I also find it very confusing that the custom type that was added using "add_custom_type" is listed twice in the tree of available nodes.

Also, I would rather see that custom types added to a scene tree do not have the original script attached. This always requires me to having to select the "extend script" option from the context menu.

**Comment by swarnimarun** [comment]
The first one of adding GDScript to the list only happens when you set a class_name and on C# it's WIP for future releases.
The second one of having multiple scripts has already been discussed to death in various issues and the answer has been No so I don't think that even makes sense anymore.

**Comment by silkentrance** [comment]
@swarnimarun and what about keeping an internal reference from nodes in your scene tree that reference custom nodes. I think that this is what is happening under the hood for built-in types.
This would be very convenient as the user will no longer be presented with the original script that implements the custom node, causing him or her to accidentally change the original source while he or she just wanted to extend upon it.

**Comment by ismslv** [comment]
@swarnimarun ,

> The first one of adding GDScript to the list only happens when you set a class_name and on C# it's WIP for future releases.

Do you think that it's normal to have nodes appearing twice in the list? Also, class_name nests and custom type does not nest at all.

> The second one of having multiple scripts has already been discussed to death in various issues and the answer has been No so I don't think that even makes sense anymore.

You have misunderstood the question. It's not about having multiple scripts, on the contrary - it's about hiding a script linked to a node of custom type. That is, to make custom types behave at least a bit more like built-in ones.

**Comment by silkentrance** [comment]
ATM when adding a custom control to the scene, you get

```
[node name="RyeCCPSection" type="Control" parent="VBoxContainer"]
margin_right = 1024.0
margin_bottom = 600.0
size_flags_horizontal = 3
size_flags_vertical = 3
script = ExtResource( 1 )
```

Here, the original script is enforced.

Why not just

```
[node name="RyeCCPSection" type="RyeCCPSection" parent="VBoxContainer"]
margin_right = 1024.0
margin_bottom = 600.0
size_flags_horizontal = 3
size_flags_vertical = 3
```

That way, I will be required to extend upon the original script which was registered by add_custom_type?

**Comment by swarnimarun** [comment]
@silkentrance In case of built-in Nodes they are not scripts but complete Scripts and we handle all the code in GDScript as Scripts cause it's not compiled to C++ but is a runtime execution.

**Comment by swarnimarun** [comment]
@ismslv Yes adding two scripts and hiding one is also not desirable, you don't really understand the situation do you? Let me put it in Layman's terms, adding any more than one Script to any Node impossible at the moment and is too much work to do otherwise, and the decision of the Devs is clearly to not get into the mess of adding such a feature.

**Comment by swarnimarun** [comment]
As for the first one, I will take another look into understanding what you mean. And Then hat it would mean for the Engine, and will discuss it further.

**Comment by silkentrance** [comment]
@swarnimarun which does not prevent us from thinking out of the box, or does it?

**Comment by silkentrance** [comment]
@swarnimarun have got any pointers on how autoload singletons and plugins will be manifested during runtime? I am currently having a hard time figuring this out. Maybe it is just in the GDScript module, which would prevent these definitions from being shared with other parts of the system, namely C#, but then again, I am just guessing.

**Comment by swarnimarun** [comment]
@silkentrance Singletons/Autoloads are quite simple actually a node is added to the top of the tree and then a script is added to it. The name is used to refer to this node/script and then you got it. :P

This is what enables you do use everything like _process, _ready and all inside a autoload. Really nice. 

I hadn't been following along Godot much for a while so I am not sure about the state of C# but C# needs marshalling to access GDScript, and usually it would need to do it through a Script object. Unless this feature(there was an issue about it iirc), was added to make it simpler.

**Comment by swarnimarun** [comment]
As for plugins it's the same Editor ones are loaded with the Editor load and custom types and all are just specific nodes with Scripts.

C# also has ability to be made into tool scripts and theoretically should work well for Editor Plugins but haven't done it so I can't elaborate on it.

**Comment by KoBeWi** [comment]
Not sure about the first one, but the second suggestion already has an issue: #6067
Especially this comment comes with a nice solution: https://github.com/godotengine/godot/issues/6067#issuecomment-469575425

**Comment by willnationsdev** [comment]
So, Godot currently has two separate "custom named script" systems: The Custom Types and the Script Classes.

Custom Types came before and added a simple icon+script auto-attachment to a type in the editor. Only the editor is aware of this system (no runtime support of the name) and scripts don't have any way to access this information at design-time (no API exposed for it). It is just a shortcut, as you've stated, with the benefit that, at the time, it was relatively easy to implement and worked generically for all scripting languages.

With 3.1, reduz added a feature to use Script Classes which allow an individual scripting language to add support for exposing a name for the script and registering it to a runtime singleton in the core. This then allows scripting languages to access the information globally, from design-time and runtime, as well as allows the editor to take advantage of it. What's more, because scripting languages are the ones that expose this data, the scripts themselves can define it without needing an EditorPlugin or some editor setting to configure the name.

Thus, reduz gave GDScript the `class_name` keyword to define a name/icon and I gave NativeScript the `script_class_name/icon` properties to define the same for NativeScript resource files. GDScript also goes the extra mile and generates global variables for all script classes.

Now, the editor's CreateDialog will generate an entry for every Custom Type and Script Class, so if you have a GDScript in an EditorPlugin that is registered as a CustomType with `add_custom_type`, and then you ALSO give that script a name via the `class_name` keyword, then you are registering it to both systems and it will show up twice. I believe this is what you are doing to see that behavior.

The truly optimized solution, imo, is to add support for Script Classes in CSharpScript, PluginScript, and VisualScript (none of which I really knew how to do at the time I was implementing most of this stuff), and then to deprecate/remove the CustomType system from Godot entirely in the 4.0 release so that there is only one named script system. That would also stand to clean up a lot of code in the engine which currently has to check both systems when handling scripted types.

---

As for the notion of having two scripts, one hidden and one overriding, I have already attempted, on numerous occasions (that horse has been beaten was past death at this point) to argue for a position like this, but reduz is adamant that a script should not and will never be hidden or made to look any different from other scripts in the editor, the argument being that users are smart and we shouldn't try to hide or manipulate truthful information from them. And on that point, he's right. The truth of the matter is, you have objects and you have scripts on them.

Even if you have a "custom type" or "script class", that script is not grafted onto the underlying class in any way. At runtime, if you just do `set_script(null)`, it reverts back to being the underlying C++ class, no matter what. I once even went through the effort of re-implementing the Object core so that it would fully simulate such double-script behaviors and the PR was rejected because of its truth-hiding nature.

**Comment by KoBeWi** [comment]
Feature and improvement proposals for the Godot Engine are now being discussed and reviewed in a dedicated [**Godot Improvement Proposals (GIP)**](https://github.com/godotengine/godot-proposals) ([godotengine/godot-proposals](https://github.com/godotengine/godot-proposals)) issue tracker. The GIP tracker has a detailed issue template designed so that proposals include all the relevant information to start a productive discussion and help the community assess the validity of the proposal for the engine.

The main ([godotengine/godot](https://github.com/godotengine/godot)) tracker is now solely dedicated to bug reports and Pull Requests, enabling contributors to have a better focus on bug fixing work. Therefore, we are now closing all older feature proposals on the main issue tracker.

If you are interested in this feature proposal, please [**open a new proposal on the GIP tracker**](https://github.com/godotengine/godot-proposals/issues/new/choose) following the given issue template (after checking that it doesn't exist already). Be sure to reference this closed issue if it includes any relevant discussion (which you are also encouraged to summarize in the new proposal). Thanks in advance!

**Comment by willnationsdev** [comment]
To be more specific, this is tracked in godotengine/godot-proposals#22 which I am working on.
